File: helix/activation.py
"""Activation codec (L4 / HELIX ①) — how a hidden-state tensor crosses the wire.

Only activations cross device boundaries in a layer-shard pipeline, and during prefill
they dominate bandwidth (``d_model × seq_len`` per hop). This module abstracts *how* an
activation is put on the wire so the encoding can improve without touching the ring
driver:

* :class:`RawActivationCodec` — exact ``float`` list. Reference / correctness tests.
* :class:`Int8ActivationCodec` — per-tensor symmetric int8 quantization (≈4× smaller
  than fp32 before framing). This is the practical, pure-Python half of HELIX ①.

The remaining half — rateless FEC (fountain/RaptorQ) so a lost datagram doesn't stall
the ring — plugs in behind the same interface and is better done in the host (native);
the interface is here so that swap needs no ring-driver change.

Payloads are JSON-embeddable (dict). A binary body path would remove the base64 overhead
of the quantized form and is a follow-up; even with it, int8 is a large net win.
"""

from __future__ import annotations

import base64
from abc import ABC, abstractmethod
from typing import Any, Dict, List


class ActivationCodec(ABC):
    @abstractmethod
    def encode(self, vec: List[float]) -> Dict[str, Any]:
        """Return a JSON-embeddable representation of ``vec``."""

    @abstractmethod
    def decode(self, payload: Dict[str, Any]) -> List[float]:
        """Inverse of :meth:`encode`."""


class RawActivationCodec(ActivationCodec):
    """Exact fp32 list — lossless, used to verify ring correctness."""

    def encode(self, vec: List[float]) -> Dict[str, Any]:
        return {"raw": [float(x) for x in vec]}

    def decode(self, payload: Dict[str, Any]) -> List[float]:
        return [float(x) for x in payload["raw"]]


class Int8ActivationCodec(ActivationCodec):
    """Per-tensor symmetric int8 quantization: ``q = round(x / scale)``, ``scale = max|x|/127``.

    Lossy (error ≤ scale/2 per element) but ~4× smaller than fp32 and far smaller than an
    fp32 JSON list. Stored as unsigned bytes (offset +128) so each element is one byte.
    """

    def encode(self, vec: List[float]) -> Dict[str, Any]:
        m = max((abs(float(x)) for x in vec), default=0.0)
        scale = m / 127.0 if m > 0 else 1.0
        q = bytes(((max(-127, min(127, int(round(x / scale)))) + 128) & 0xFF) for x in vec)
        return {"q": base64.b64encode(q).decode("ascii"), "s": scale, "n": len(vec)}

    def decode(self, payload: Dict[str, Any]) -> List[float]:
        raw = base64.b64decode(payload["q"])
        scale = float(payload["s"])
        return [(b - 128) * scale for b in raw]

File: integration/chatterui_llamacpp/js/activation.mjs
// HELIX activation codec — JS reference (helix/activation.py), for the ChatterUI *Level 3* /
// Option B ring: how a hidden-state tensor crosses a HELIX frame.
//
//   RawActivationCodec  — exact fp32 list (correctness; used by the ring proof).
//   Int8ActivationCodec — per-tensor symmetric int8 (HELIX ①, ~4x smaller). Pinned by
//                         helix/spec/vectors.json (activation_codec) and ./conformance.mjs.
//
// The int8 quantizer mirrors Python's round-half-to-even so the bytes match exactly.

// round half to even (Python's round()), so JS and Python quantize identically.
function rintEven(x) {
  const f = Math.floor(x);
  const d = x - f;
  if (d < 0.5) return f;
  if (d > 0.5) return f + 1;
  return f % 2 === 0 ? f : f + 1;
}

export class RawActivationCodec {
  encode(vec) { return { raw: vec.map((x) => x) }; }
  decode(payload) { return payload.raw.map((x) => x); }
}

export class Int8ActivationCodec {
  encode(vec) {
    let m = 0;
    for (const x of vec) m = Math.max(m, Math.abs(x));
    const scale = m > 0 ? m / 127 : 1.0;
    const q = Buffer.alloc(vec.length);
    for (let i = 0; i < vec.length; i++) {
      const c = Math.max(-127, Math.min(127, rintEven(vec[i] / scale)));
      q[i] = (c + 128) & 0xff;
    }
    return { q: q.toString("base64"), s: scale, n: vec.length };
  }

  decode(payload) {
    const raw = Buffer.from(payload.q, "base64");
    const scale = payload.s;
    return Array.from(raw, (b) => (b - 128) * scale);
  }
}
